fix evalpoly type instability and 0-length case

stevengj opened the pull request:
A few fixes/improvements to the `evalpoly` function:

1. Fixes #56699: returns zero for an empty tuple or array of coefficients
2. Fixes a type-instability in `evalpoly(x, ::Vector)`
3. Adds a missing `@inbounds` annotation
4. Trivial: Renames a local variable `ex` that was-copy-pasted from the macro version (where `ex` referred to an expression) but makes no sense here for the numeric sum.
5. Removes some extraneous methods (by tightening a method signature, as suggested by @nsajko) — should still dispatch to the same implementations as before.

Probably should be backported.

stevengj committed:
fix evalpoly type instability and 0-length case

stevengj committed:
more evalpoly tests

LilithHafner commented on base/math.jl:
```diff
@@ -103,17 +103,20 @@ function evalpoly(x, p::Tuple)
-    ex = p[end]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
+    s = oftype(one(x) * p0, p0)
+    N <= 1 && return s
     for i in N-1:-1:1
```
```suggestion
    for i in N-1:-1:1
```
If `N <= 1` then `N-1:-1:1` is empty.

stevengj committed:
rm unnecessary test

nsajko commented on base/math.jl:
```diff
@@ -142,16 +144,17 @@ function evalpoly(z::Complex, p::Tuple)
     N = length(p)
-    a = p[end]
-    b = p[end-1]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
```
```suggestion
    p0 = iszero(N) ? reduce_empty_iter(+, p) : @inbounds p[N]
```

nsajko commented on base/math.jl:
```diff
@@ -103,17 +103,19 @@ function evalpoly(x, p::Tuple)
     N = length(p)
-    ex = p[end]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
+    s = oftype(one(x) * p0, p0)
```
This construction doesn't seem to guarantee any useful behavior. I think you're assuming that `one(x) isa typeof(x)` is guaranteed for any reasonable `x`? That's not actually part of the interface/contract of `one`, so this doesn't help with type stability in general.

Perhaps something like this would be better (although it'd still cause unnecessary computation when the arithmetic can't be optimized out by the compiler):

```suggestion
    s = oftype(muladd(x, p0, p0), p0)
```

stevengj replied:
`one` seems like it is the correct function here to get the desired behavior with unitful types (where `evalpoly` on an array of coefficients is inherently type-unstable, however).

`x * p0` or `muladd(x, p0, p0)` are both incorrect for unitful `x`.

For non-unitful types, the documentation of `one(x)` says that "If possible, one(x) returns a value of the same type as x".

nsajko commented on base/math.jl:
```diff
@@ -142,16 +144,17 @@ function evalpoly(z::Complex, p::Tuple)
-    a = p[end]
-    b = p[end-1]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
+    N <= 1 && return oftype(one(z) * p0, p0)
```
As above, this doesn't help with type stability in general.

stevengj replied:
As I explained above, `one` is the correct function to handle unitful types.

nsajko reviewed (changes requested):
I'm not sure if opening this PR was polite, considering a newcomer submitted #56699 trying to fix the same issue, and the issue is marked as "good first issue". In any case, that PR is now closed, so here are some comments.

stevengj commented on base/math.jl:
```diff
@@ -103,17 +103,19 @@ function evalpoly(x, p::Tuple)
     Base.require_one_based_indexing(p)
     N = length(p)
-    ex = p[end]
+    @inbounds p0 = N == 0 ? reduce_empty_iter(+, p) : p[N]
```
Yes, there is a bounds check for `p[N]`  without `@inbouds` here.

In my mind, applying `@inbounds` to the entire line is *easier* to read.   And it doesn't affect any of the functions here except for `p[N]`.

Regarding `iszero`, `N = length(p)` should be an `Int`, so `iszero` has no particular advantage here.  (I'm also not sure `iszero` is actually faster for `BigInt`, by the way, since `BigInt(N) == 0` ends up calling a specialized method — it doesn't promote `0` to `BigInt`.

stevengj commented on base/math.jl:
```diff
@@ -103,17 +103,19 @@ function evalpoly(x, p::Tuple)
+        @inbounds s = muladd(x, s, p[i])
     end
-    ex
+    return s
```
I could, but touching this code is a good time for a cleanup — the variable name `ex` makes no sense here, it was clearly copy-pasted from the macro version of `evalpoly` where `ex` is an expression.

If we don't fix it now, I'm unlikely to go through the whole PR process just to fix a badly named local variable.

stevengj commented on base/math.jl:
```diff
@@ -103,17 +103,19 @@ function evalpoly(x, p::Tuple)
+    s = oftype(one(x) * p0, p0)
     for i in N-1:-1:1
-        ex = muladd(x, ex, p[i])
+        @inbounds s = muladd(x, s, p[i])
```
Like I said above, I find it a lot more readable if we don't interleave `@inbounds` with the expression, and it has the same effect either way.

stevengj committed:
Update base/math.jl

Co-authored-by: Neven Sajko <s@purelymail.com>

stevengj committed:
Update base/math.jl

Co-authored-by: Neven Sajko <s@purelymail.com>

stevengj committed:
Update base/math.jl

Co-authored-by: Neven Sajko <s@purelymail.com>

stevengj committed:
Update base/math.jl

Co-authored-by: Neven Sajko <s@purelymail.com>

LilithHafner commented on base/math.jl:
```diff
@@ -142,16 +144,17 @@ function evalpoly(z::Complex, p::Tuple)
 end
 evalpoly(z::Complex, p::Tuple{<:Any}) = p[1]
-
+evalpoly(z::Complex, ::Tuple{}) = zero(one(z)) # dimensionless zero, i.e. 0 * z^0
```
If we're going do demote `0.0 + 0.0im` to `0.0`, might as well demote it all the way to `false`.

stevengj committed:
rm extraneous methods

stevengj commented:
CI failures look unrelated.

MasonProtter commented on base/math.jl:
```diff
@@ -103,20 +103,23 @@ function evalpoly(x, p::Tuple)
+    s = oftype(one(x) * p0, p0)
     for i in N-1:-1:1
-        ex = muladd(x, ex, p[i])
+        s = muladd(x, s, @inbounds p[i])
```
Since this is a generic function, this is not necessarily safe. It's entirely possible that `muladd(x, y, z)` does some  internal indexing of some array, and the `@inbounds` annotation on the outside would then infect the implementation of `muladd(x, y, z)`. 

An alternative would be to write
```julia
pi = @inbounds p[i]
muladd(x, s, pi)
```

stevengj commented:
Yay, green.  Good to merge?

oscardssmith commented:
why is triage added here?

stevengj commented:
Because it's been ready to merge for 2 months and needs someone to review it?  Is there a better label?

LilithHafner reviewed (commented):
This makes the existing use of `@generated` on line 95 invalid because the two branches now return different types:

```julia-repl
julia> evalpoly(1.0, (1,))
1

julia> Base.Math._evalpoly(1.0, (1,))
1.0

julia> @less evalpoly(1, (1,))
function evalpoly(x, p::Tuple)
    if @generated
        N = length(p.parameters::Core.SimpleVector)
        ex = :(p[end])
        for i in N-1:-1:1
            ex = :(muladd(x, $ex, p[$i]))
        end
        ex
    else
        _evalpoly(x, p)
    end
end
```

LilithHafner commented on base/math.jl:
```diff
@@ -103,20 +103,23 @@ function evalpoly(x, p::Tuple)
         _evalpoly(x, p)
     end
 end
+evalpoly(x, ::Tuple{}) = zero(one(x)) # dimensionless zero, i.e. 0 * x^0
```
I agree that this is right. Would you please add a test that will fail under the `zero(x)` implementation?

stevengj replied:
added a test

stevengj committed:
added test for empty-coefficients case, with units

stevengj commented:
Oh ye gods, børked rebase.

giordano commented:
Git history is messed up now.

stevengj commented:
Should be fixed now.

stevengj committed:
ensure that different generated brnches return same type

stevengj commented:
@LilithHafner, I fixed the issue with the `@generated` branches you noted in https://github.com/JuliaLang/julia/pull/56707#pullrequestreview-2526890685, and added a test.

stevengj commented:
Currently getting:
```
 Test threw exception
  Expression: #= /cache/build/tester-amdci4-13/julialang/julia-master/julia-f7b9e88614/share/julia/stdlib/v1.12/LinearAlgebra/test/matmul.jl:647 =# @evalpoly(A33, 1.0I, 1.0I) == I + A33
  MethodError: Cannot `convert` an object of type LinearAlgebra.UniformScaling{Float64} to an object of type Matrix{ComplexF64}
  The function `convert` exists, but no method is defined for this combination of argument types.
```
This could be fixed by replacing `oftype(one(x) * p[end], p[end])` with simply `one(x) * p[end]`.  However, in that case I'm concerned that https://github.com/JuliaLang/LinearAlgebra.jl/issues/1161 will get worse — it will then start silently giving the wrong answer for matrix `x` and scalar coefficients, because of the weird behavior of `muladd` (https://github.com/JuliaLang/LinearAlgebra.jl/issues/1162) in this case.

So maybe this PR needs to wait for LinearAlgebra.jl to have a separate PR that fixes https://github.com/JuliaLang/LinearAlgebra.jl/issues/1161?

LilithHafner commented:
This now breaks the identity `evalpoly(x, (y,)) === y`. e.g:

```julia
julia> evalpoly(:sploo, (:bork,))
:bork # now throws

julia> evalpoly(1.0, (true,))::Bool
true # now 1.0

julia> evalpoly(rand(2,2), (1,))
1 # This should possibly be Float64[1 0; 0 1] now throws
```

According to the docstring, `evalpoly(x, (y,))` should return `x^0 * y` which, after this PR, in the first two cases it does and in the third case it hits JuliaLang/LinearAlgebra.jl#1161.

I think that "breakage" is a bugfix.

stevengj commented:
> This now breaks the identity `evalpoly(x, (y,)) === y`

Not sure why this identity should be expected from the definition?

LilithHafner commented:
> Not sure why this identity should be expected from the definition?

If you write out a polynomial in the form `ax^2+bx+c`, rather than `a_2x^2+a_1x^1+a_0x^0`, then the degree 0 case is just `a` which would imply the result would be egal to `a`.

> I guess you agree with the new behavior?

Yes, I do. Regardless of the above.

fingolfin commented:
There were CI failures, e.g.
```
mith-C07ZM05NJYVY-0/julialang/julia-master/julia-f7b9e88614/share/julia/stdlib/v1.12/LinearAlgebra/test/matmul.jl:647
  Test threw exception
  Expression: #= /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-grannysmith-C07ZM05NJYVY.0/build/default-grannysmith
-C07ZM05NJYVY-0/julialang/julia-master/julia-f7b9e88614/share/julia/stdlib/v1.12/LinearAlgebra/test/matmul.jl:647 =# @evalpoly(A33, 1.0I, 1.0I) == I + A33
  MethodError: Cannot `convert` an object of type LinearAlgebra.UniformScaling{Float64} to an object of type Matrix{ComplexF64}
  The function `convert` exists, but no method is defined for this combination of argument types.
  Closest candidates are:
    Matrix{T}(::LinearAlgebra.UniformScaling, !Matched::Tuple{Int64, Int64}) where T
     @ LinearAlgebra /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-grannysmith-C07ZM05NJYVY.0/build/default-grannys
mith-C07ZM05NJYVY-0/julialang/julia-master/julia-f7b9e88614/share/julia/stdlib/v1.12/LinearAlgebra/src/uniformscaling.jl:423
    Matrix{T}(::LinearAlgebra.UniformScaling, !Matched::Integer, !Matched::Integer) where T
     @ LinearAlgebra /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-grannysmith-C07ZM05NJYVY.0/build/default-grannys
mith-C07ZM05NJYVY-0/julialang/julia-master/julia-f7b9e88614/share/julia/stdlib/v1.12/LinearAlgebra/src/uniformscaling.jl:431
    convert(::Type{T}, !Matched::AbstractArray) where T<:Array
     @ Base array.jl:614
```

stevengj commented:
Blocked by https://github.com/JuliaLang/LinearAlgebra.jl/issues/1161 as noted above.

MasonProtter commented on base/math.jl:
```diff
@@ -94,7 +94,7 @@ julia> evalpoly(2, (1, 2, 3))
     if @generated
         N = length(p.parameters::Core.SimpleVector)
-        ex = :(p[end])
+        ex = :(oftype(one(x) * p[end], p[end]))
```
I worry a bit about this causing performance regressions (especially in cases where multiplication is not cheap), but I also get that it's likely important. 

Ideally we'd be able to do something in the type domain here rather than an actual multiplication, but oh well i guess.

stevengj replied:
Hopefully the compiler can optimize out `one(x) * p[end]` in most cases, and just compute the type, since the result is not used.

fingolfin replied:
Only if the multiplication is simple enough and free of side effects like memory allocation? If the scalar is eg a BigInt it may not be able to optimize this away?
